### DISCUSSION TITLE
Capture value from response in one yaml and pass it onto another yaml

### DISCUSSION THREAD

**Discussion by priya-dsp** [artilleryio/artillery#2491] [Artillery]
Hi, we have created multiple Artillery scripts - yaml's for different scenarios . We need to capture a value from response of an API request in first Yaml and pass this value onto another Artillery script- yaml. Second yaml is supposed to read the values captured from first yaml. How to achieve this ?

**Comment by hassy** [comment]
hi @priya-dsp 👋 All variables (including ones that are captured from a response) are local to that specific virtual user. There's no way to pass them between VUs. If you need to pre-load data of some sort into a running Artillery test there's a couple of different things you could do. What's your use case?

**Reply by priya-dsp** [reply]
For example: we are capturing access token from an API in first yaml and we need to pass this dynamic token into second yaml which have api's that are dependent on token. We have used afterResponsehook to call js function and capture token value and assigned into new js variable. when we tried to call that js variable into another yaml, its not working. I assume, JS function called using afterResponse is also local to that first yaml. Is there an option to call JS function globally that works for all yamls?

**Reply by bernardobridge** [reply]
Hi @priya-dsp 👋 

You're trying to use Artillery in a way that it's not currently designed to work. As far as I understand you're orchestrating multiple Artillery tests (yamls) together using something like a bash script - each of those will be an isolated NodeJS process which won't be able to share resources with others. If you want to share information between different executions, you'd have to find a way to share a stateful component between them (e.g. write the token to a file that the next test can pick up, or a shared database).

--- 

The native way to accomplish that sort of thing with Artillery is to create reusable patterns. For example, let's say you need an API token in your tests. Maybe you turn that into a reusable javascript function that can be imported into the processor of any of the tests. For example, something like:

```js
const axios = require('axios');

async function getAccessToken (context, events, done) { 
    const response = await axios.post('https://placeholder.url/api/token', {});

    const accessToken = response.data.access_token;
    context.vars['accessToken'] = accessToken;
    done();
};
```

Then, each of the different YAMLs that needs an access token from the API, can call that in a [`before` hook ](https://www.artillery.io/docs/reference/test-script#before-and-after-sections), and use that token for the duration of the test.

**Reply by priya-dsp** [reply]
Hi @bernardobridge .Thanks for your input. I tried this Javascript and called it in Artillery script using hook but i'm getting [Error]: Cannot read properties of undefined (reading 'vars')  at getAccessToken and  [Error]: done is not a function. Is there anything missing ?

**Reply by bernardobridge** [reply]
Hi @priya-dsp. I'd need details in order to try to help - could you please share the Artillery YAML and the NodeJS code you are using and getting an error with?

**Reply by priya-dsp** [reply]
Hi @bernardobridge , it's working. Got it fixed. Thanks for your help. Have used function hook in yaml to call JS file. Now, we are able to read accesstoken from javascript to yaml. 

`const axios = require('axios');    
 
async function getAccessToken (context, events, done) { 
    const url = "https://login.xyz";
    const postdata = {
	grant_type: "client_credentials",
        client_id: "clientid_value"
	};
   const headers = {
         'Content-Type': "application/form-urlencoded"	
	};
    const response = await axios.post(url,postdata,{headers});
    console.log(response.data.access_token);
    accessToken = response.data.access_token;  
    context.vars['pAccessToken'] = accessToken;   
    done();  
}`

**Comment by priya-dsp** [comment]
Hi @bernardobridge , with above approach i am able to create a token and call the js function in each module.
But when i call js function its creating new token everytime for each module.
Requirement is i need to create token only once and reuse the same token in different modules (Each module is a seperate yaml).

In jmeter we have property function to pass the variables from 1 thread group to another thread group. Is there any function similar to jmeter in artillery ? Is there any other approach apart from writing token into a file so that it should work with multiple users ?
